File: speech/Record/for_testing/Record/Class1.java
/*
 * MATLAB Compiler: 6.4 (R2017a)
 * Date: Sat Oct 14 22:03:44 2017
 * Arguments: 
 * "-B""macro_default""-W""java:Record,Class1""-T""link:lib""-d""C:\\Users\\Fatih 
 * Furkan\\Documents\\MATLAB\\speech\\Record\\for_testing""class{Class1:C:\\Users\\Fatih 
 * Furkan\\Documents\\MATLAB\\speech\\Record.m,C:\\Users\\Fatih 
 * Furkan\\Documents\\MATLAB\\speech\\test_word.m,C:\\Users\\Fatih 
 * Furkan\\Documents\\MATLAB\\speech\\train_all_words.m,C:\\Users\\Fatih 
 * Furkan\\Documents\\MATLAB\\speech\\train_word.m}""-a""C:\\Users\\Fatih 
 * Furkan\\Documents\\MATLAB\\speech\\hmm_model_files\\aþaðý.mat""-a""C:\\Users\\Fatih 
 * Furkan\\Documents\\MATLAB\\speech\\hmm_model_files\\aþaðý1.mat""-a""C:\\Users\\Fatih 
 * Furkan\\Documents\\MATLAB\\speech\\hmm_model_files\\saða.mat""-a""C:\\Users\\Fatih 
 * Furkan\\Documents\\MATLAB\\speech\\hmm_model_files\\saða1.mat""-a""C:\\Users\\Fatih 
 * Furkan\\Documents\\MATLAB\\speech\\hmm_model_files\\sola.mat""-a""C:\\Users\\Fatih 
 * Furkan\\Documents\\MATLAB\\speech\\hmm_model_files\\sola1.mat""-a""C:\\Users\\Fatih 
 * Furkan\\Documents\\MATLAB\\speech\\hmm_model_files\\týkla.mat""-a""C:\\Users\\Fatih 
 * Furkan\\Documents\\MATLAB\\speech\\hmm_model_files\\týkla1.mat""-a""C:\\Users\\Fatih 
 * Furkan\\Documents\\MATLAB\\speech\\hmm_model_files\\yukarý.mat""-a""C:\\Users\\Fatih 
 * Furkan\\Documents\\MATLAB\\speech\\hmm_model_files\\yukarý1.mat"
 */

package Record;

import com.mathworks.toolbox.javabuilder.*;
import com.mathworks.toolbox.javabuilder.internal.*;
import java.util.*;

/**
 * The <code>Class1</code> class provides a Java interface to MATLAB functions. 
 * The interface is compiled from the following files:
 * <pre>
 *  C:\Users\Fatih Furkan\Documents\MATLAB\speech\Record.m
 *  C:\Users\Fatih Furkan\Documents\MATLAB\speech\test_word.m
 *  C:\Users\Fatih Furkan\Documents\MATLAB\speech\train_all_words.m
 *  C:\Users\Fatih Furkan\Documents\MATLAB\speech\train_word.m
 * </pre>
 * The {@link #dispose} method <b>must</b> be called on a <code>Class1</code> instance 
 * when it is no longer needed to ensure that native resources allocated by this class 
 * are properly freed.
 * @version 0.0
 */
public class Class1 extends MWComponentInstance<Class1>
{
    /**
     * Tracks all instances of this class to ensure their dispose method is
     * called on shutdown.
     */
    private static final Set<Disposable> sInstances = new HashSet<Disposable>();

    /**
     * Maintains information used in calling the <code>Record</code> MATLAB function.
     */
    private static final MWFunctionSignature sRecordSignature =
        new MWFunctionSignature(/* max outputs = */ 0,
                                /* has varargout = */ false,
                                /* function name = */ "Record",
                                /* max inputs = */ 0,
                                /* has varargin = */ false);
    /**
     * Maintains information used in calling the <code>test_word</code> MATLAB function.
     */
    private static final MWFunctionSignature sTest_wordSignature =
        new MWFunctionSignature(/* max outputs = */ 1,
                                /* has varargout = */ false,
                                /* function name = */ "test_word",
                                /* max inputs = */ 0,
                                /* has varargin = */ false);
    /**
     * Maintains information used in calling the <code>train_all_words</code> MATLAB 
     *function.
     */
    private static final MWFunctionSignature sTrain_all_wordsSignature =
        new MWFunctionSignature(/* max outputs = */ 0,
                                /* has varargout = */ false,
                                /* function name = */ "train_all_words",
                                /* max inputs = */ 0,
                                /* has varargin = */ false);
    /**
     * Maintains information used in calling the <code>train_word</code> MATLAB function.
     */
    private static final MWFunctionSignature sTrain_wordSignature =
        new MWFunctionSignature(/* max outputs = */ 0,
                                /* has varargout = */ false,
                                /* function name = */ "train_word",
                                /* max inputs = */ 2,
                                /* has varargin = */ false);

    /**
     * Shared initialization implementation - private
     * @throws MWException An error has occurred during the function call.
     */
    private Class1 (final MWMCR mcr) throws MWException
    {
        super(mcr);
        // add this to sInstances
        synchronized(Class1.class) {
            sInstances.add(this);
        }
    }

    /**
     * Constructs a new instance of the <code>Class1</code> class.
     * @throws MWException An error has occurred during the function call.
     */
    public Class1() throws MWException
    {
        this(RecordMCRFactory.newInstance());
    }
    
    private static MWComponentOptions getPathToComponentOptions(String path)
    {
        MWComponentOptions options = new MWComponentOptions(new MWCtfExtractLocation(path),
                                                            new MWCtfDirectorySource(path));
        return options;
    }
    
    /**
     * @deprecated Please use the constructor {@link #Class1(MWComponentOptions componentOptions)}.
     * The <code>com.mathworks.toolbox.javabuilder.MWComponentOptions</code> class provides an API to set the
     * path to the component.
     * @param pathToComponent Path to component directory.
     * @throws MWException An error has occurred during the function call.
     */
    public Class1(String pathToComponent) throws MWException
    {
        this(RecordMCRFactory.newInstance(getPathToComponentOptions(pathToComponent)));
    }
    
    /**
     * Constructs a new instance of the <code>Class1</code> class. Use this constructor 
     * to specify the options required to instantiate this component.  The options will 
     * be specific to the instance of this component being created.
     * @param componentOptions Options specific to the component.
     * @throws MWException An error has occurred during the function call.
     */
    public Class1(MWComponentOptions componentOptions) throws MWException
    {
        this(RecordMCRFactory.newInstance(componentOptions));
    }
    
    /** Frees native resources associated with this object */
    public void dispose()
    {
        try {
            super.dispose();
        } finally {
            synchronized(Class1.class) {
                sInstances.remove(this);
            }
        }
    }
  
    /**
     * Invokes the first MATLAB function specified to MCC, with any arguments given on
     * the command line, and prints the result.
     *
     * @param args arguments to the function
     */
    public static void main (String[] args)
    {
        try {
            MWMCR mcr = RecordMCRFactory.newInstance();
            mcr.runMain( sRecordSignature, args);
            mcr.dispose();
        } catch (Throwable t) {
            t.printStackTrace();
        }
    }
    
    /**
     * Calls dispose method for each outstanding instance of this class.
     */
    public static void disposeAllInstances()
    {
        synchronized(Class1.class) {
            for (Disposable i : sInstances) i.dispose();
            sInstances.clear();
        }
    }

    /**
     * Provides the interface for calling the <code>Record</code> MATLAB function 
     * where the first argument, an instance of List, receives the output of the MATLAB function and
     * the second argument, also an instance of List, provides the input to the MATLAB function.
     * <p>
     * Description as provided by the author of the MATLAB function:
     * </p>
     * <pre>
     * % get(recObj)
     * %recObj = audiorecorder;
     * </pre>
     * @param lhs List in which to return outputs. Number of outputs (nargout) is
     * determined by allocated size of this List. Outputs are returned as
     * sub-classes of <code>com.mathworks.toolbox.javabuilder.MWArray</code>.
     * Each output array should be freed by calling its <code>dispose()</code>
     * method.
     *
     * @param rhs List containing inputs. Number of inputs (nargin) is determined
     * by the allocated size of this List. Input arguments may be passed as
     * sub-classes of <code>com.mathworks.toolbox.javabuilder.MWArray</code>, or
     * as arrays of any supported Java type. Arguments passed as Java types are
     * converted to MATLAB arrays according to default conversion rules.
     * @throws MWException An error has occurred during the function call.
     */
    public void Record(List lhs, List rhs) throws MWException
    {
        fMCR.invoke(lhs, rhs, sRecordSignature);
    }

    /**
     * Provides the interface for calling the <code>Record</code> MATLAB function 
     * where the first argument, an Object array, receives the output of the MATLAB function and
     * the second argument, also an Object array, provides the input to the MATLAB function.
     * <p>
     * Description as provided by the author of the MATLAB function:
     * </p>
     * <pre>
     * % get(recObj)
     * %recObj = audiorecorder;
     * </pre>
     * @param lhs array in which to return outputs. Number of outputs (nargout)
     * is determined by allocated size of this array. Outputs are returned as
     * sub-classes of <code>com.mathworks.toolbox.javabuilder.MWArray</code>.
     * Each output array should be freed by calling its <code>dispose()</code>
     * method.
     *
     * @param rhs array containing inputs. Number of inputs (nargin) is
     * determined by the allocated size of this array. Input arguments may be
     * passed as sub-classes of
     * <code>com.mathworks.toolbox.javabuilder.MWArray</code>, or as arrays of
     * any supported Java type. Arguments passed as Java types are converted to
     * MATLAB arrays according to default conversion rules.
     * @throws MWException An error has occurred during the function call.
     */
    public void Record(Object[] lhs, Object[] rhs) throws MWException
    {
        fMCR.invoke(Arrays.asList(lhs), Arrays.asList(rhs), sRecordSignature);
    }

    /**
     * Provides the standard interface for calling the <code>Record</code> MATLAB function with 
     * 0 input arguments.
     * Input arguments may be passed as sub-classes of
     * <code>com.mathworks.toolbox.javabuilder.MWArray</code>, or as arrays of
     * any supported Java type. Arguments passed as Java types are converted to
     * MATLAB arrays according to default conversion rules.
     *
     * <p>
     * Description as provided by the author of the MATLAB function:
     * </p>
     * <pre>
     * % get(recObj)
     * %recObj = audiorecorder;
     * </pre>
     * @param rhs The inputs to the MATLAB function.
     * @return Array of length nargout containing the function outputs. Outputs
     * are returned as sub-classes of
     * <code>com.mathworks.toolbox.javabuilder.MWArray</code>. Each output array
     * should be freed by calling its <code>dispose()</code> method.
     * @throws MWException An error has occurred during the function call.
     */
    public Object[] Record(Object... rhs) throws MWException
    {
        Object[] lhs = new Object[0];
        fMCR.invoke(Arrays.asList(lhs), 
                    MWMCR.getRhsCompat(rhs, sRecordSignature), 
                    sRecordSignature);
        return lhs;
    }
    /**
     * Provides the interface for calling the <code>test_word</code> MATLAB function 
     * where the first argument, an instance of List, receives the output of the MATLAB function and
     * the second argument, also an instance of List, provides the input to the MATLAB function.
     * <p>
     * Description as provided by the author of the MATLAB function:
     * </p>
     * <pre>
     * % List of words we have already trained and have
     * % HMM model files in hmm_model_files directory.
     * </pre>
     * @param lhs List in which to return outputs. Number of outputs (nargout) is
     * determined by allocated size of this List. Outputs are returned as
     * sub-classes of <code>com.mathworks.toolbox.javabuilder.MWArray</code>.
     * Each output array should be freed by calling its <code>dispose()</code>
     * method.
     *
     * @param rhs List containing inputs. Number of inputs (nargin) is determined
     * by the allocated size of this List. Input arguments may be passed as
     * sub-classes of <code>com.mathworks.toolbox.javabuilder.MWArray</code>, or
     * as arrays of any supported Java type. Arguments passed as Java types are
     * converted to MATLAB arrays according to default conversion rules.
     * @throws MWException An error has occurred during the function call.
     */
    public void test_word(List lhs, List rhs) throws MWException
    {
        fMCR.invoke(lhs, rhs, sTest_wordSignature);
    }

    /**
     * Provides the interface for calling the <code>test_word</code> MATLAB function 
     * where the first argument, an Object array, receives the output of the MATLAB function and
     * the second argument, also an Object array, provides the input to the MATLAB function.
     * <p>
     * Description as provided by the author of the MATLAB function:
     * </p>
     * <pre>
     * % List of words we have already trained and have
     * % HMM model files in hmm_model_files directory.
     * </pre>
     * @param lhs array in which to return outputs. Number of outputs (nargout)
     * is determined by allocated size of this array. Outputs are returned as
     * sub-classes of <code>com.mathworks.toolbox.javabuilder.MWArray</code>.
     * Each output array should be freed by calling its <code>dispose()</code>
     * method.
     *
     * @param rhs array containing inputs. Number of inputs (nargin) is
     * determined by the allocated size of this array. Input arguments may be
     * passed as sub-classes of
     * <code>com.mathworks.toolbox.javabuilder.MWArray</code>, or as arrays of
     * any supported Java type. Arguments passed as Java types are converted to
     * MATLAB arrays according to default conversion rules.
     * @throws MWException An error has occurred during the function call.
     */
    public void test_word(Object[] lhs, Object[] rhs) throws MWException
    {
        fMCR.invoke(Arrays.asList(lhs), Arrays.asList(rhs), sTest_wordSignature);
    }

    /**
     * Provides the standard interface for calling the <code>test_word</code> MATLAB function with 
     * 0 input arguments.
     * Input arguments may be passed as sub-classes of
     * <code>com.mathworks.toolbox.javabuilder.MWArray</code>, or as arrays of
     * any supported Java type. Arguments passed as Java types are converted to
     * MATLAB arrays according to default conversion rules.
     *
     * <p>
     * Description as provided by the author of the MATLAB function:
     * </p>
     * <pre>
     * % List of words we have already trained and have
     * % HMM model files in hmm_model_files directory.
     * </pre>
     * @param nargout Number of outputs to return.
     * @param rhs The inputs to the MATLAB function.
     * @return Array of length nargout containing the function outputs. Outputs
     * are returned as sub-classes of
     * <code>com.mathworks.toolbox.javabuilder.MWArray</code>. Each output array
     * should be freed by calling its <code>dispose()</code> method.
     * @throws MWException An error has occurred during the function call.
     */
    public Object[] test_word(int nargout, Object... rhs) throws MWException
    {
        Object[] lhs = new Object[nargout];
        fMCR.invoke(Arrays.asList(lhs), 
                    MWMCR.getRhsCompat(rhs, sTest_wordSignature), 
                    sTest_wordSignature);
        return lhs;
    }
    /**
     * Provides the interface for calling the <code>train_all_words</code> MATLAB function 
     * where the first argument, an instance of List, receives the output of the MATLAB function and
     * the second argument, also an instance of List, provides the input to the MATLAB function.
     * @param lhs List in which to return outputs. Number of outputs (nargout) is
     * determined by allocated size of this List. Outputs are returned as
     * sub-classes of <code>com.mathworks.toolbox.javabuilder.MWArray</code>.
     * Each output array should be freed by calling its <code>dispose()</code>
     * method.
     *
     * @param rhs List containing inputs. Number of inputs (nargin) is determined
     * by the allocated size of this List. Input arguments may be passed as
     * sub-classes of <code>com.mathworks.toolbox.javabuilder.MWArray</code>, or
     * as arrays of any supported Java type. Arguments passed as Java types are
     * converted to MATLAB arrays according to default conversion rules.
     * @throws MWException An error has occurred during the function call.
     */
    public void train_all_words(List lhs, List rhs) throws MWException
    {
        fMCR.invoke(lhs, rhs, sTrain_all_wordsSignature);
    }

    /**
     * Provides the interface for calling the <code>train_all_words</code> MATLAB function 
     * where the first argument, an Object array, receives the output of the MATLAB function and
     * the second argument, also an Object array, provides the input to the MATLAB function.
     * @param lhs array in which to return outputs. Number of outputs (nargout)
     * is determined by allocated size of this array. Outputs are returned as
     * sub-classes of <code>com.mathworks.toolbox.javabuilder.MWArray</code>.
     * Each output array should be freed by calling its <code>dispose()</code>
     * method.
     *
     * @param rhs array containing inputs. Number of inputs (nargin) is
     * determined by the allocated size of this array. Input arguments may be
     * passed as sub-classes of
     * <code>com.mathworks.toolbox.javabuilder.MWArray</code>, or as arrays of
     * any supported Java type. Arguments passed as Java types are converted to
     * MATLAB arrays according to default conversion rules.
     * @throws MWException An error has occurred during the function call.
     */
    public void train_all_words(Object[] lhs, Object[] rhs) throws MWException
    {
        fMCR.invoke(Arrays.asList(lhs), Arrays.asList(rhs), sTrain_all_wordsSignature);
    }

    /**
     * Provides the standard interface for calling the <code>train_all_words</code> MATLAB function with 
     * 0 input arguments.
     * Input arguments may be passed as sub-classes of
     * <code>com.mathworks.toolbox.javabuilder.MWArray</code>, or as arrays of
     * any supported Java type. Arguments passed as Java types are converted to
     * MATLAB arrays according to default conversion rules.
     *
     * @param rhs The inputs to the MATLAB function.
     * @return Array of length nargout containing the function outputs. Outputs
     * are returned as sub-classes of
     * <code>com.mathworks.toolbox.javabuilder.MWArray</code>. Each output array
     * should be freed by calling its <code>dispose()</code> method.
     * @throws MWException An error has occurred during the function call.
     */
    public Object[] train_all_words(Object... rhs) throws MWException
    {
        Object[] lhs = new Object[0];
        fMCR.invoke(Arrays.asList(lhs), 
                    MWMCR.getRhsCompat(rhs, sTrain_all_wordsSignature), 
                    sTrain_all_wordsSignature);
        return lhs;
    }
    /**
     * Provides the interface for calling the <code>train_word</code> MATLAB function 
     * where the first argument, an instance of List, receives the output of the MATLAB function and
     * the second argument, also an instance of List, provides the input to the MATLAB function.
     * <p>
     * Description as provided by the author of the MATLAB function:
     * </p>
     * <pre>
     * % read wav file and generate cepstrum
     * </pre>
     * @param lhs List in which to return outputs. Number of outputs (nargout) is
     * determined by allocated size of this List. Outputs are returned as
     * sub-classes of <code>com.mathworks.toolbox.javabuilder.MWArray</code>.
     * Each output array should be freed by calling its <code>dispose()</code>
     * method.
     *
     * @param rhs List containing inputs. Number of inputs (nargin) is determined
     * by the allocated size of this List. Input arguments may be passed as
     * sub-classes of <code>com.mathworks.toolbox.javabuilder.MWArray</code>, or
     * as arrays of any supported Java type. Arguments passed as Java types are
     * converted to MATLAB arrays according to default conversion rules.
     * @throws MWException An error has occurred during the function call.
     */
    public void train_word(List lhs, List rhs) throws MWException
    {
        fMCR.invoke(lhs, rhs, sTrain_wordSignature);
    }

    /**
     * Provides the interface for calling the <code>train_word</code> MATLAB function 
     * where the first argument, an Object array, receives the output of the MATLAB function and
     * the second argument, also an Object array, provides the input to the MATLAB function.
     * <p>
     * Description as provided by the author of the MATLAB function:
     * </p>
     * <pre>
     * % read wav file and generate cepstrum
     * </pre>
     * @param lhs array in which to return outputs. Number of outputs (nargout)
     * is determined by allocated size of this array. Outputs are returned as
     * sub-classes of <code>com.mathworks.toolbox.javabuilder.MWArray</code>.
     * Each output array should be freed by calling its <code>dispose()</code>
     * method.
     *
     * @param rhs array containing inputs. Number of inputs (nargin) is
     * determined by the allocated size of this array. Input arguments may be
     * passed as sub-classes of
     * <code>com.mathworks.toolbox.javabuilder.MWArray</code>, or as arrays of
     * any supported Java type. Arguments passed as Java types are converted to
     * MATLAB arrays according to default conversion rules.
     * @throws MWException An error has occurred during the function call.
     */
    public void train_word(Object[] lhs, Object[] rhs) throws MWException
    {
        fMCR.invoke(Arrays.asList(lhs), Arrays.asList(rhs), sTrain_wordSignature);
    }

    /**
     * Provides the standard interface for calling the <code>train_word</code> MATLAB function with 
     * 2 comma-separated input arguments.
     * Input arguments may be passed as sub-classes of
     * <code>com.mathworks.toolbox.javabuilder.MWArray</code>, or as arrays of
     * any supported Java type. Arguments passed as Java types are converted to
     * MATLAB arrays according to default conversion rules.
     *
     * <p>
     * Description as provided by the author of the MATLAB function:
     * </p>
     * <pre>
     * % read wav file and generate cepstrum
     * </pre>
     * @param rhs The inputs to the MATLAB function.
     * @return Array of length nargout containing the function outputs. Outputs
     * are returned as sub-classes of
     * <code>com.mathworks.toolbox.javabuilder.MWArray</code>. Each output array
     * should be freed by calling its <code>dispose()</code> method.
     * @throws MWException An error has occurred during the function call.
     */
    public Object[] train_word(Object... rhs) throws MWException
    {
        Object[] lhs = new Object[0];
        fMCR.invoke(Arrays.asList(lhs), 
                    MWMCR.getRhsCompat(rhs, sTrain_wordSignature), 
                    sTrain_wordSignature);
        return lhs;
    }
}

File: speech/Record1/for_testing/Class1.cs
/*
* MATLAB Compiler: 6.4 (R2017a)
* Date: Sat Oct 14 19:53:12 2017
* Arguments:
* "-B""macro_default""-W""dotnet:Record1,Class1,4.0,private""-T""link:lib""-d""C:\Users\Fa
* tih Furkan\Documents\MATLAB\speech\Record1\for_testing""-v""class{Class1:C:\Users\Fatih
* Furkan\Documents\MATLAB\speech\Record.m,C:\Users\Fatih
* Furkan\Documents\MATLAB\speech\test_word.m,C:\Users\Fatih
* Furkan\Documents\MATLAB\speech\train_all_words.m,C:\Users\Fatih
* Furkan\Documents\MATLAB\speech\train_word.m}""-a""C:\Users\Fatih
* Furkan\Documents\MATLAB\speech\hmm_model_files\aþaðý.mat""-a""C:\Users\Fatih
* Furkan\Documents\MATLAB\speech\hmm_model_files\aþaðý1.mat""-a""C:\Users\Fatih
* Furkan\Documents\MATLAB\speech\hmm_model_files\saða.mat""-a""C:\Users\Fatih
* Furkan\Documents\MATLAB\speech\hmm_model_files\saða1.mat""-a""C:\Users\Fatih
* Furkan\Documents\MATLAB\speech\hmm_model_files\sola.mat""-a""C:\Users\Fatih
* Furkan\Documents\MATLAB\speech\hmm_model_files\sola1.mat""-a""C:\Users\Fatih
* Furkan\Documents\MATLAB\speech\hmm_model_files\týkla.mat""-a""C:\Users\Fatih
* Furkan\Documents\MATLAB\speech\hmm_model_files\týkla1.mat""-a""C:\Users\Fatih
* Furkan\Documents\MATLAB\speech\hmm_model_files\yukarý.mat""-a""C:\Users\Fatih
* Furkan\Documents\MATLAB\speech\hmm_model_files\yukarý1.mat"
*/
using System;
using System.Reflection;
using System.IO;
using MathWorks.MATLAB.NET.Arrays;
using MathWorks.MATLAB.NET.Utility;

#if SHARED
[assembly: System.Reflection.AssemblyKeyFile(@"")]
#endif

namespace Record1
{

  /// <summary>
  /// The Class1 class provides a CLS compliant, MWArray interface to the MATLAB
  /// functions contained in the files:
  /// <newpara></newpara>
  /// C:\Users\Fatih Furkan\Documents\MATLAB\speech\Record.m
  /// <newpara></newpara>
  /// C:\Users\Fatih Furkan\Documents\MATLAB\speech\test_word.m
  /// <newpara></newpara>
  /// C:\Users\Fatih Furkan\Documents\MATLAB\speech\train_all_words.m
  /// <newpara></newpara>
  /// C:\Users\Fatih Furkan\Documents\MATLAB\speech\train_word.m
  /// </summary>
  /// <remarks>
  /// @Version 4.0
  /// </remarks>
  public class Class1 : IDisposable
  {
    #region Constructors

    /// <summary internal= "true">
    /// The static constructor instantiates and initializes the MATLAB Runtime instance.
    /// </summary>
    static Class1()
    {
      if (MWMCR.MCRAppInitialized)
      {
        try
        {
          Assembly assembly= Assembly.GetExecutingAssembly();

          string ctfFilePath= assembly.Location;

          int lastDelimiter= ctfFilePath.LastIndexOf(@"\");

          ctfFilePath= ctfFilePath.Remove(lastDelimiter, (ctfFilePath.Length - lastDelimiter));

          string ctfFileName = "Record1.ctf";

          Stream embeddedCtfStream = null;

          String[] resourceStrings = assembly.GetManifestResourceNames();

          foreach (String name in resourceStrings)
          {
            if (name.Contains(ctfFileName))
            {
              embeddedCtfStream = assembly.GetManifestResourceStream(name);
              break;
            }
          }
          mcr= new MWMCR("",
                         ctfFilePath, embeddedCtfStream, true);
        }
        catch(Exception ex)
        {
          ex_ = new Exception("MWArray assembly failed to be initialized", ex);
        }
      }
      else
      {
        ex_ = new ApplicationException("MWArray assembly could not be initialized");
      }
    }


    /// <summary>
    /// Constructs a new instance of the Class1 class.
    /// </summary>
    public Class1()
    {
      if(ex_ != null)
      {
        throw ex_;
      }
    }


    #endregion Constructors

    #region Finalize

    /// <summary internal= "true">
    /// Class destructor called by the CLR garbage collector.
    /// </summary>
    ~Class1()
    {
      Dispose(false);
    }


    /// <summary>
    /// Frees the native resources associated with this object
    /// </summary>
    public void Dispose()
    {
      Dispose(true);

      GC.SuppressFinalize(this);
    }


    /// <summary internal= "true">
    /// Internal dispose function
    /// </summary>
    protected virtual void Dispose(bool disposing)
    {
      if (!disposed)
      {
        disposed= true;

        if (disposing)
        {
          // Free managed resources;
        }

        // Free native resources
      }
    }


    #endregion Finalize

    #region Methods

    /// <summary>
    /// Provides a void output, 0-input MWArrayinterface to the Record MATLAB function.
    /// </summary>
    /// <remarks>
    /// M-Documentation:
    /// get(recObj)
    /// recObj = audiorecorder;
    /// </remarks>
    ///
    public void Record()
    {
      mcr.EvaluateFunction(0, "Record", new MWArray[]{});
    }


    /// <summary>
    /// Provides the standard 0-input MWArray interface to the Record MATLAB function.
    /// </summary>
    /// <remarks>
    /// M-Documentation:
    /// get(recObj)
    /// recObj = audiorecorder;
    /// </remarks>
    /// <param name="numArgsOut">The number of output arguments to return.</param>
    /// <returns>An Array of length "numArgsOut" containing the output
    /// arguments.</returns>
    ///
    public MWArray[] Record(int numArgsOut)
    {
      return mcr.EvaluateFunction(numArgsOut, "Record", new MWArray[]{});
    }


    /// <summary>
    /// Provides a void output, 0-input MWArrayinterface to the test_word MATLAB
    /// function.
    /// </summary>
    /// <remarks>
    /// M-Documentation:
    /// List of words we have already trained and have
    /// HMM model files in hmm_model_files directory.
    /// </remarks>
    ///
    public void test_word()
    {
      mcr.EvaluateFunction(0, "test_word", new MWArray[]{});
    }


    /// <summary>
    /// Provides the standard 0-input MWArray interface to the test_word MATLAB function.
    /// </summary>
    /// <remarks>
    /// M-Documentation:
    /// List of words we have already trained and have
    /// HMM model files in hmm_model_files directory.
    /// </remarks>
    /// <param name="numArgsOut">The number of output arguments to return.</param>
    /// <returns>An Array of length "numArgsOut" containing the output
    /// arguments.</returns>
    ///
    public MWArray[] test_word(int numArgsOut)
    {
      return mcr.EvaluateFunction(numArgsOut, "test_word", new MWArray[]{});
    }


    /// <summary>
    /// Provides a void output, 0-input MWArrayinterface to the train_all_words MATLAB
    /// function.
    /// </summary>
    /// <remarks>
    /// </remarks>
    ///
    public void train_all_words()
    {
      mcr.EvaluateFunction(0, "train_all_words", new MWArray[]{});
    }


    /// <summary>
    /// Provides the standard 0-input MWArray interface to the train_all_words MATLAB
    /// function.
    /// </summary>
    /// <remarks>
    /// </remarks>
    /// <param name="numArgsOut">The number of output arguments to return.</param>
    /// <returns>An Array of length "numArgsOut" containing the output
    /// arguments.</returns>
    ///
    public MWArray[] train_all_words(int numArgsOut)
    {
      return mcr.EvaluateFunction(numArgsOut, "train_all_words", new MWArray[]{});
    }


    /// <summary>
    /// Provides a void output, 0-input MWArrayinterface to the train_word MATLAB
    /// function.
    /// </summary>
    /// <remarks>
    /// M-Documentation:
    /// read wav file and generate cepstrum
    /// </remarks>
    ///
    public void train_word()
    {
      mcr.EvaluateFunction(0, "train_word", new MWArray[]{});
    }


    /// <summary>
    /// Provides a void output, 1-input MWArrayinterface to the train_word MATLAB
    /// function.
    /// </summary>
    /// <remarks>
    /// M-Documentation:
    /// read wav file and generate cepstrum
    /// </remarks>
    /// <param name="wordName">Input argument #1</param>
    ///
    public void train_word(MWArray wordName)
    {
      mcr.EvaluateFunction(0, "train_word", wordName);
    }


    /// <summary>
    /// Provides a void output, 2-input MWArrayinterface to the train_word MATLAB
    /// function.
    /// </summary>
    /// <remarks>
    /// M-Documentation:
    /// read wav file and generate cepstrum
    /// </remarks>
    /// <param name="wordName">Input argument #1</param>
    /// <param name="numStates">Input argument #2</param>
    ///
    public void train_word(MWArray wordName, MWArray numStates)
    {
      mcr.EvaluateFunction(0, "train_word", wordName, numStates);
    }


    /// <summary>
    /// Provides the standard 0-input MWArray interface to the train_word MATLAB
    /// function.
    /// </summary>
    /// <remarks>
    /// M-Documentation:
    /// read wav file and generate cepstrum
    /// </remarks>
    /// <param name="numArgsOut">The number of output arguments to return.</param>
    /// <returns>An Array of length "numArgsOut" containing the output
    /// arguments.</returns>
    ///
    public MWArray[] train_word(int numArgsOut)
    {
      return mcr.EvaluateFunction(numArgsOut, "train_word", new MWArray[]{});
    }


    /// <summary>
    /// Provides the standard 1-input MWArray interface to the train_word MATLAB
    /// function.
    /// </summary>
    /// <remarks>
    /// M-Documentation:
    /// read wav file and generate cepstrum
    /// </remarks>
    /// <param name="numArgsOut">The number of output arguments to return.</param>
    /// <param name="wordName">Input argument #1</param>
    /// <returns>An Array of length "numArgsOut" containing the output
    /// arguments.</returns>
    ///
    public MWArray[] train_word(int numArgsOut, MWArray wordName)
    {
      return mcr.EvaluateFunction(numArgsOut, "train_word", wordName);
    }


    /// <summary>
    /// Provides the standard 2-input MWArray interface to the train_word MATLAB
    /// function.
    /// </summary>
    /// <remarks>
    /// M-Documentation:
    /// read wav file and generate cepstrum
    /// </remarks>
    /// <param name="numArgsOut">The number of output arguments to return.</param>
    /// <param name="wordName">Input argument #1</param>
    /// <param name="numStates">Input argument #2</param>
    /// <returns>An Array of length "numArgsOut" containing the output
    /// arguments.</returns>
    ///
    public MWArray[] train_word(int numArgsOut, MWArray wordName, MWArray numStates)
    {
      return mcr.EvaluateFunction(numArgsOut, "train_word", wordName, numStates);
    }



    /// <summary>
    /// This method will cause a MATLAB figure window to behave as a modal dialog box.
    /// The method will not return until all the figure windows associated with this
    /// component have been closed.
    /// </summary>
    /// <remarks>
    /// An application should only call this method when required to keep the
    /// MATLAB figure window from disappearing.  Other techniques, such as calling
    /// Console.ReadLine() from the application should be considered where
    /// possible.</remarks>
    ///
    public void WaitForFiguresToDie()
    {
      mcr.WaitForFiguresToDie();
    }



    #endregion Methods

    #region Class Members

    private static MWMCR mcr= null;

    private static Exception ex_= null;

    private bool disposed= false;

    #endregion Class Members
  }
}
